File: lib/daraja.ts
const SANDBOX_OAUTH = "https://sandbox.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials";
const PRODUCTION_OAUTH = "https://api.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials";
const SANDBOX_STK = "https://sandbox.safaricom.co.ke/mpesa/stkpush/v1/processrequest";
const PRODUCTION_STK = "https://api.safaricom.co.ke/mpesa/stkpush/v1/processrequest";

export async function getOAuthToken(
  consumerKey: string,
  consumerSecret: string,
  env: string
): Promise<string> {
  const auth = Buffer.from(`${consumerKey}:${consumerSecret}`).toString("base64");
  const url = env === "production" ? PRODUCTION_OAUTH : SANDBOX_OAUTH;

  const res = await fetch(url, {
    method: "GET",
    headers: { Authorization: `Basic ${auth}` },
  });

  const data = await res.json();
  if (!data.access_token) {
    throw new Error(data.error_description || "Failed to get OAuth token");
  }
  return data.access_token as string;
}

function formatTimestamp(): string {
  const now = new Date();
  const y = now.getFullYear();
  const M = String(now.getMonth() + 1).padStart(2, "0");
  const d = String(now.getDate()).padStart(2, "0");
  const h = String(now.getHours()).padStart(2, "0");
  const m = String(now.getMinutes()).padStart(2, "0");
  const s = String(now.getSeconds()).padStart(2, "0");
  return `${y}${M}${d}${h}${m}${s}`;
}

function normalizePhone(phone: string): string {
  let p = phone.replace(/\s+/g, "");
  if (p.startsWith("+254")) p = "254" + p.slice(4);
  else if (p.startsWith("0")) p = "254" + p.slice(1);
  else if (!p.startsWith("254")) p = "254" + p;
  return p;
}

export interface DarajaConfig {
  consumerKey: string;
  consumerSecret: string;
  passkey: string;
  shortcode: string;
  callbackUrl: string;
  env: string;
}

export async function sendSTKPush(
  config: DarajaConfig,
  phone: string,
  amount: number,
  reference: string
) {
  const token = await getOAuthToken(config.consumerKey, config.consumerSecret, config.env);
  const timestamp = formatTimestamp();
  const password = Buffer.from(
    `${config.shortcode}${config.passkey}${timestamp}`
  ).toString("base64");
  const normalizedPhone = normalizePhone(phone);
  const url = config.env === "production" ? PRODUCTION_STK : SANDBOX_STK;

  const payload = {
    BusinessShortCode: config.shortcode,
    Password: password,
    Timestamp: timestamp,
    TransactionType: "CustomerPayBillOnline",
    Amount: Math.floor(amount),
    PartyA: normalizedPhone,
    PartyB: config.shortcode,
    PhoneNumber: normalizedPhone,
    CallBackURL: config.callbackUrl,
    AccountReference: reference,
    TransactionDesc: reference,
  };

  const res = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  return res.json();
}

File: backend/internal/daraja/daraja.go
package daraja

import (
	"bytes"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const sandboxOAuth = "https://sandbox.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials"
const productionOAuth = "https://api.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials"
const sandboxSTK = "https://sandbox.safaricom.co.ke/mpesa/stkpush/v1/processrequest"
const productionSTK = "https://api.safaricom.co.ke/mpesa/stkpush/v1/processrequest"

type Config struct {
	ConsumerKey    string
	ConsumerSecret string
	Passkey        string
	Shortcode      string
	CallbackURL    string
	Env            string
}

type Client struct {
	httpClient *http.Client
}

func New() *Client {
	return &Client{
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *Client) GetOAuthToken(cfg Config) (string, error) {
	auth := base64.StdEncoding.EncodeToString([]byte(cfg.ConsumerKey + ":" + cfg.ConsumerSecret))
	url := sandboxOAuth
	if cfg.Env == "production" {
		url = productionOAuth
	}

	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return "", fmt.Errorf("create oauth request: %w", err)
	}
	req.Header.Set("Authorization", "Basic "+auth)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("oauth request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("read oauth response: %w", err)
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return "", fmt.Errorf("parse oauth response (status %d): %s", resp.StatusCode, string(body))
	}

	token, ok := data["access_token"].(string)
	if !ok || token == "" {
		errDesc, _ := data["error_description"].(string)
		return "", fmt.Errorf("oauth failed: %s", errDesc)
	}

	return token, nil
}

func formatTimestamp() string {
	now := time.Now()
	return fmt.Sprintf("%04d%02d%02d%02d%02d%02d",
		now.Year(), now.Month(), now.Day(),
		now.Hour(), now.Minute(), now.Second())
}

func normalizePhone(phone string) string {
	p := strings.ReplaceAll(phone, " ", "")
	if strings.HasPrefix(p, "+254") {
		p = "254" + p[4:]
	} else if strings.HasPrefix(p, "0") {
		p = "254" + p[1:]
	} else if !strings.HasPrefix(p, "254") {
		p = "254" + p
	}
	return p
}

type STKPushResult struct {
	ResponseCode        string `json:"ResponseCode"`
	ResponseDescription string `json:"ResponseDescription"`
	CheckoutRequestID   string `json:"CheckoutRequestID"`
	MerchantRequestID   string `json:"MerchantRequestID"`
	ErrorMessage        string `json:"errorMessage,omitempty"`
}

func (c *Client) SendSTKPush(cfg Config, phone string, amount float64, reference string) (*STKPushResult, error) {
	token, err := c.GetOAuthToken(cfg)
	if err != nil {
		return nil, fmt.Errorf("get token: %w", err)
	}

	timestamp := formatTimestamp()
	password := base64.StdEncoding.EncodeToString([]byte(cfg.Shortcode + cfg.Passkey + timestamp))
	normalizedPhone := normalizePhone(phone)

	url := sandboxSTK
	if cfg.Env == "production" {
		url = productionSTK
	}

	payload := map[string]interface{}{
		"BusinessShortCode": cfg.Shortcode,
		"Password":          password,
		"Timestamp":         timestamp,
		"TransactionType":   "CustomerPayBillOnline",
		"Amount":            int(amount),
		"PartyA":            normalizedPhone,
		"PartyB":            cfg.Shortcode,
		"PhoneNumber":       normalizedPhone,
		"CallBackURL":       cfg.CallbackURL,
		"AccountReference":  reference,
		"TransactionDesc":   reference,
	}

	body, _ := json.Marshal(payload)
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create stk request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("stk request: %w", err)
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	var result STKPushResult
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("parse stk response (status %d): %s", resp.StatusCode, string(respBody))
	}

	return &result, nil
}

func GenerateToken() string {
	b := make([]byte, 24)
	rand.Read(b)
	return "mash_live_" + base64.RawURLEncoding.EncodeToString(b)
}

var safaricomIPs = map[string]bool{
	"196.201.214.200": true,
	"196.201.214.206": true,
	"196.201.213.200": true,
	"196.201.213.206": true,
	"196.201.214.208": true,
	"196.201.213.208": true,
}

func IsValidSafaricomIP(ip string) bool {
	return safaricomIPs[ip]
}
